feat: document Isaac Lab x AICA System bridge (#303)
aica-technology/api: docs/docs/examples/guides/isaaclab-aica-bridge.md

---

> sidebar_position: 6

Reviewer: ```suggestion
sidebar_position: 11
```

With Isaac Lab as simulator, users can build scenes in Isaac Lab, command simulated robots using AICA System, validate
performance, switch the hardware interface to a real robot, and hit play with no code changes required.

## Installing Isaac Lab

Begin by cloning AICA's fork of
[Isaac Lab](https://github.com/aica-technology/isaac-lab/tree/f17384a1b487630128b4782ce02166565ef4464f/).

Once the repository is cloned, build and start the Docker container by running:

```shell
python3 docker/container.py start
```

Next, enter the running container using:

```shell
python3 docker/container.py enter
```

This ensures you are inside a development environment where Isaac Lab and all required dependencies are already
installed. Run the following command in the same terminal to verify the installation:

```shell
python3 scripts/custom/aica_bridge/run_bridge.py --scene basic_scene
```

This will spawn a UR5e robot, a ground plane, and lights. If you see the UR5e robot in the scene, then the installation
was successful and you are ready to proceed with the next steps.

If not, carefully go over the instructions again or reach out to AICA for help.

## Connecting AICA System to Isaac Lab

> In this section, we’ll demonstrate a simple example of using a point attractor to move a simulated UR5e robot in Isaac Lab toward a target frame, and then manipulate that frame within the 3D visualization of AICA Studio. If you haven’t already, please follow the [Point Attractor Example](../core-components/point-attractor) to create an AICA application that moves a robot’s end-effector using a point attractor. We’ll use that application as the foundation for this guide.

Reviewer: ```suggestion
In this section, the Point Attractor example from [this page](../core-components/point-attractor.md) is adapted to
control a simulated UR5e robot in Isaac Lab. Follow this example before proceeding here, as it serves as the foundation
for the next steps. 
```

is **1803**, and it must match the `ft_sensor_port` in the hardware interface configuration.  
- **joint_names**: Lists the joint names from the URDF that AICA will command. For example, if you are using a Franka Panda robot with a gripper 
but only want to control the arm, you can specify:  
  `"panda_joint1", "panda_joint2", "panda_joint3", "panda_joint4", "panda_joint5", "panda_joint6", "panda_joint7"`.
  The simulator will then only send states and accept commands for those joints.  If you want to control all joints, you can keep the default `"*"`.
- **command_interface**: Defines the command type accepted by the simulator. The default is `"positions"`, but you can set 
it to `"velocities"` or `"torques"` as needed. If a mismatched command type is received, the simulator will stop with a `ValueError`.  
- **headless**: When set to `true`, runs the simulator in headless mode (without a user interface), useful for remote simulations or running the simulation at high frequencies.  
- **device**: Specifies the compute device for the simulation. The default is `"cuda"` for GPU acceleration, but you can switch to `"cpu"` if GPU resources are unavailable.  

> 

Reviewer 2: ```suggestion
```

Ensure these parameters are correctly configured to enable seamless communication between the simulator and your AICA
application. In case you want to run the simulator with different parameters, you can do so by running the following
command in the `run_bridge.py` script:

```shell
python3 scripts/custom/aica_bridge/run_bridge.py --scene <your_scene_name> --rate <simulation_rate> --force_sensor <force_sensor_name_in_urdf> --state_port <state_port> --command_port <command_port> --force_port <force_port> --joint_names <comma_separated_joint_names_to_control> --command_interface <positions/velocities/torques> --headless <true/false> --device <cuda/cpu>
```

> 

Reviewer 2: ```suggestion
```

### Running the AICA Application

> When the simulator is running, you can execute your AICA application by first chosing the UR5e URDF file that you just

Reviewer 2: ```suggestion
When the simulator is running, you can execute your AICA application by first choosing the UR5e URDF file that you just
```

```

This will spawn a UR5e robot, a ground plane, and lights. If you see the UR5e robot in the scene, then the installation
was successful and you are ready to proceed with the next steps.

If not, carefully go over the instructions again or reach out to AICA for help.

## Connecting AICA System to Isaac Lab

In this section, we’ll walk through a simple example of using a point attractor to move a UR5e simulated robot in Isaac
Lab to a target frame and then manipulate that frame in the 3D visualization of the AICA Studio.

### Creating a New Scene in Isaac Lab

To create a new scene, you should define a scene configuration class that inherits from `InteractiveSceneCfg`. Various
examples of scene config classes can be found in the
[scenes](https://github.com/aica-technology/isaac-lab/blob/f17384a1b487630128b4782ce02166565ef4464f/scripts/custom/aica_bridge/scenes)
directory of the Isaac Lab repository.

Note that the scene lives entirely in Isaac Lab and the definitions of the assets used in the scene should be defined
there. The 3D visualization in the AICA Studio will only display the robot and mirror the robot's movements, but it will
not display the scene itself.

> Once you've defined your scene configuration class, register it by adding a corresponding key to the `scenes` dictionary

Reviewer 3: This is the first reference to a key - to me it's not clear what it is or how to register it

---
sidebar_position: 6
title: A guide for using Isaac Lab as a simulator for the AICA System
---

import application from './assets/isaaclab-aica-bridge-application.png'

# A guide for using Isaac Lab as a simulator for AICA System

## Motivation

By interfacing the AICA System with Isaac Lab, we establish a workflow for developing, testing, and deploying robotic
applications. This connection provides several key benefits:

1. **RL Policy Testing**: AICA’s RL Policy Component SDK allows developers to deploy Reinforcement Learning (RL) models
   directly onto real hardware through components. These models can be trained in Isaac Lab, and with the AICA System
   interacting directly with Isaac Lab, users can validate trained policies under the same conditions in which they
   were learned.

2. **Reliable Policy Validation**: Developers can monitor the behavior of the trained policies and test the effect of
   various parameters, enabling confident transitions from simulation to real-world deployment.

> 3. **Digital Twin Control**: Beyond RL, running the AICA System with Isaac Lab provides users with ways to interact with

Author: In my understanding, a digital twin is a digital version (i.e. simulated version) of a physical entity. In a sense, that a robot in simulation that closely resembles the real robot in terms of mass, structure, interactions in a physics simulated world is a digital twin.

> 3. **Digital Twin Control**: Beyond RL, running the AICA System with Isaac Lab provides users with ways to interact with

Reviewer: Yeah I guess it depends who you ask, Gemini says "a virtual replica of a physical robot or robotic system, mirroring its behavior and state in real-time through sensor data" and Nvidia defines it in your way: https://www.nvidia.com/en-us/glossary/digital-twin/

   digital twins of their robots. Applications can be authored, tested, and validated entirely in simulation before
   connecting to actual hardware. This improves safety and enables rapid iteration in early stages, helping streamline
   the overall development cycle.

With Isaac Lab as simulator, users can build complete automation pipelines in Isaac Lab, interact with them using AICA

> System, validate performance, switch the hardware interface to a real robot, and hit play with no code changes required.

Author: I do see your point here, yes. I changed to control robot using AICA System

```shell
python3 docker/container.py start
```

This command will build the Docker image and start the container in the background.

Next, enter the running container using:

```shell
python3 docker/container.py enter
```

This ensures you are inside a development environment where Isaac Lab and all required dependencies are already
installed.

Once done, verify the installation by running

```shell
python3 scripts/custom/aica_bridge/run_bridge.py --scene basic_scene
```

This will spawn a UR5e robot, a ground plane and lights. If you see the UR5e robot in the scene, then the installation
was successful and you are ready to proceed with the next steps.

If not, carefully go over the instructions again or reach out to our [support email](contact@aica.tech) for help.

# Connecting AICA System to Isaac Lab

In this section, we’ll walk through a simple example of using a point attractor to move a UR5e simulated robot in Isaac
Lab to a target frame and then manipulate that frame in the 3D visualization of AICA Studio.

## Creating a New Scene in Isaac Lab

To create a new scene, you should define a scene config class that inherits from InteractiveSceneCfg. Various examples

> of scene config classes can be found in the [`scenes`](https://github.com/aica-technology/isaac-lab/blob/main/scripts/custom/aica_bridge/scenes) directory of the Isaac Lab

Reviewer 2: for the links, I think we should reference a tag on that repo (which means we should also create one), otherwise we risk breaking the docs by making changes upstream

  expects position commands. You can change this to `"velocity"` if required. If a mismatched command type is received,
  the simulator will terminate with a `ValueError` indicating the mismatch.
- **headless**: When set to `true`, the simulator runs in headless mode, which is useful for running simulations without
  a graphical interface.
- **device**: Specifies the device to use for simulation. The default is `cuda`, which utilizes GPU acceleration. You
  can change this to `cpu` if you want to run the simulation on the CPU instead.

Ensure these parameters are correctly configured to enable seamless communication between the simulator and your AICA
application. In case you want to run the simulator with different parameters, you can do so by running the following
command in the `run_bridge.py` script:

```shell
python3 scripts/custom/aica_bridge/run_bridge.py --scene <your_scene_name> --rate <simulation_rate> --end_effector <end_effector_link_name> --force_sensor <true/false> --state_port <state_port> --command_port <command_port> --force_port <force_port> --command_interface <position/velocity> --headless <true/false> --device <cuda/cpu>
```

## Creating an AICA Application

> To get started, follow these steps to create a new application in **AICA Studio**:
> 

Reviewer: ```suggestion
```

simulators. In this example, we’ll use it to connect with the Isaac Lab simulator via ZMQ sockets.

To set up the hardware interface, follow these steps:

1. Go to the **Hardware** tab in **AICA Studio**.
2. Click on the Universal Robot 5e URDF to open it.
3. Click **Save As** to create a copy of the URDF file for the UR5e robot in **AICA Studio**.
4. In the **hardware** tag, replace the existing plugin with the lightweight interface plugin shown above.

> 5. Remove any unused tags, such as GPIO or other unnecessary hardware interfaces, from the URDF file.

Author: That's true, we can remove it all together.

   - **IK Velocity Controller**: Calculates the inverse kinematics for the UR5e, enabling smooth motion toward the
     target frame.

4. Connect the events and signals as shown in the image. Make sure to auto-configure and auto-activate all components
   and controllers.

<div class="text--center">
  <img src={application} alt="Signal Point Attractor Application" />
</div>

5. Press **Play**, then switch to the **3D Visualizer** tab in **AICA Studio**. You should see the UR5e robot in the
   scene. Use the **Record Frame** button to capture a target frame at the robot's tool, specify the tool frame name
   (for UR robots from the collection, that is `ur_tool0`), and then move the frame to your desired target position.
6. Once the target frame is set, parametrize the **Frame to Signal** component with the name of the target frame you
   just created.

Now that you have set up the application, let's go over the necessary steps that needs to be done on the Hardware
Interface level to connect **AICA System** to the Isaac Lab simulator.

## Configuring the Hardware Interface

> The **Hardware Interface** serves as the communication bridge between the **AICA System** and external hardware or

Reviewer: I still think this needs to come before we create the application, because the screenshot above already assumes this URDF is present

```shell
python3 scripts/custom/aica_bridge/run_bridge.py --scene <your_scene_name>
```

In this example we will run the `basic_scene` scene, which is already registered in the `scenes` dictionary.

### Running the Isaac Lab Simulator

The simulator provides a list of parameters that you should understand before configuring it:

:::tip

Some of these parameters have counterparts defined in the hardware interface used within AICA Studio and will be
introduced below.

:::

- **scene**: Specifies which scene to load in the simulator.
- **rate**: Sets the simulation update frequency in hertz (Hz). The default is **100 Hz**, but you can adjust this value
  based on your application’s requirements.
- **ft_sensor_name**: If provided, a sensor will be attached to the end-effector link.
  :::caution
  If a `<sensor>` plugin in the URDF of the hardware interface used in AICA Studio is configured, the name of the sensor
  here needs to correspond to the name of the sensor in the URDF.
  :::
- **ip_address**: Indicates the IP address of the machine running AICA Core. If the simulator and AICA Core are on the

>   same network, keep the default `"*"`.

Reviewer: Still think that we could just write `*` and avoid the quotes but TBD

---

import application from './assets/isaaclab-aica-bridge-application.png'

# A guide for using Isaac Lab as a simulator for AICA System

By interfacing the AICA System with Isaac Lab, we establish a workflow for developing, testing, and deploying robotic
applications. This connection provides several key benefits:

1. **RL Policy Testing**: AICA’s RL Policy Component SDK allows developers to deploy Reinforcement Learning (RL) models
   directly onto real hardware through components. These models can be trained in Isaac Lab, and with the AICA System
   interacting directly with Isaac Lab, users can validate trained policies under the same conditions in which they were
   learned.

2. **Physics-Based Evaluation**: Running the AICA System within a physics-based simulation such as Isaac Lab allows
   developers to observe how control algorithms respond to realistic dynamics, friction, collisions, and sensor noise.
   This ensures that behaviors tested in simulation mirror real-world performance, reducing the risk of unexpected
   failures during deployment and enabling safer, more reliable policy tuning before engaging with physical robots.

> 3. **Digital Twin Control**: Beyond RL, running the AICA System with Isaac Lab provides users with ways to interact with

Reviewer: I like points 1 and 2 much more, nice iteration. Point 3 is still a stretch for me but why not

[Isaac Lab](https://github.com/aica-technology/isaac-lab/tree/f17384a1b487630128b4782ce02166565ef4464f/).

Once the repository is cloned, build and start the Docker container by running:

```shell
python3 docker/container.py start
```

Next, enter the running container using:

```shell
python3 docker/container.py enter
```

This ensures you are inside a development environment where Isaac Lab and all required dependencies are already
installed. Run the following command in the same terminal to verify the installation:

```shell
python3 scripts/custom/aica_bridge/run_bridge.py --scene basic_scene
```

This will spawn a UR5e robot, a ground plane, and lights. If you see the UR5e robot in the scene, then the installation
was successful and you are ready to proceed with the next steps.

If not, carefully go over the instructions again or reach out to AICA for help.

## Connecting AICA System to Isaac Lab

In this section, we’ll demonstrate a simple example of using a point attractor to move a simulated UR5e robot in Isaac
Lab toward a target frame, and then manipulate that frame within the 3D visualization of AICA Studio.

First, let’s review the key steps involved in this setup:

> 1. **Creating a New Scene in Isaac Lab**: Define a scene configuration class that inherits from `InteractiveSceneCfg`

Reviewer: ```suggestion
1. **Creating a new scene in Isaac Lab**: Define a scene configuration class that inherits from `InteractiveSceneCfg`
```

   and register it in the `scenes` dictionary. This scene will include the robot model and any other objects you want to
   interact with.
2. **Running the Isaac Lab Simulator**: Launch the simulator with parameters that specify the scene, rate, force sensor,
   ports, joint names, command interface, and device.
3. **Configuring the AICA Application**: Set up a hardware interface in AICA Studio that uses the `LightWeightInterface`
   plugin to connect to the simulator.
4. **Running the AICA Application**: Start the AICA application to control the robot in the simulator.

> ### Creating a New Scene in Isaac Lab

Reviewer: ```suggestion
### Creating a new scene in Isaac Lab
```

- **force_port**: The port used to stream force/torque measurements from the simulator to the hardware interface in AICA
  Studio. The default is **1803**, and it must match the `ft_sensor_port` in the hardware interface configuration.
- **joint_names**: Lists the joint names from the URDF that AICA will command. For example, if you are using a Franka
  Panda robot with a gripper but only want to control the arm, you can specify:  
   `"panda_joint1", "panda_joint2", "panda_joint3", "panda_joint4", "panda_joint5", "panda_joint6", "panda_joint7"`. The
  simulator will then only send states and accept commands for those joints. If you want to control all joints, you can keep
  the default `"*"`.
- **command_interface**: Defines the command type accepted by the simulator. The default is `"positions"`, but you can
  set it to `"velocities"` or `"torques"` as needed. If a mismatched command type is received, the simulator will stop
  with a `ValueError`.
- **headless**: When set to `true`, runs the simulator in headless mode (without a user interface), useful for remote
  simulations or running the simulation at high frequencies.

> - **device**: Specifies the compute device for the simulation. The default is `"cuda"` for GPU acceleration, but you can

Reviewer: can we use "gpu" or does isaac lab use "cuda"?

- **force_port**: The port used to stream force/torque measurements from the simulator to the hardware interface in AICA
  Studio. The default is **1803**, and it must match the `ft_sensor_port` in the hardware interface configuration.
- **joint_names**: Lists the joint names from the URDF that AICA will command. For example, if you are using a Franka
  Panda robot with a gripper but only want to control the arm, you can specify:  
   `"panda_joint1", "panda_joint2", "panda_joint3", "panda_joint4", "panda_joint5", "panda_joint6", "panda_joint7"`. The
  simulator will then only send states and accept commands for those joints. If you want to control all joints, you can keep
  the default `"*"`.
- **command_interface**: Defines the command type accepted by the simulator. The default is `"positions"`, but you can
  set it to `"velocities"` or `"torques"` as needed. If a mismatched command type is received, the simulator will stop
  with a `ValueError`.
- **headless**: When set to `true`, runs the simulator in headless mode (without a user interface), useful for remote
  simulations or running the simulation at high frequencies.

> - **device**: Specifies the compute device for the simulation. The default is `"cuda"` for GPU acceleration, but you can

Reviewer: Ok then we can probably remove the `"` from all those inline code blocks

  switch to `"cpu"` if GPU resources are unavailable.

Ensure these parameters are correctly configured to enable seamless communication between the simulator and your AICA
application. In case you want to run the simulator with different parameters, you can do so by running the following
command in the `run_bridge.py` script:

```shell
python3 scripts/custom/aica_bridge/run_bridge.py \
  --scene <your_scene_name> \
  --rate <simulation_rate> \
  --force_sensor <force_sensor_name_in_urdf> \
  --state_port <state_port> \
  --command_port <command_port> \
  --force_port <force_port> \
  --joint_names <comma_separated_joint_names_to_control> \
  --command_interface <positions/velocities/torques> \
  --headless <true/false> \
  --device <cuda/cpu>
```

> 

Author: yep indeed, I'll write it down after this ^

1. Go to the **Hardware** tab in **AICA Studio**.
2. Click on the Universal Robot 5e URDF to open it.
3. Click **Save As** to create a copy of the URDF file for the UR5e robot in **AICA Studio**.
4. In the **hardware** tag, replace the existing plugin with the lightweight interface plugin shown above.
5. Remove any unused tags, such as GPIO or other unnecessary hardware interfaces, from the URDF file.
6. Save your changes.
7. In the **Hardware Interface** component on the graph view, select the modified URDF file for the UR5e robot from the
   drop down menu in the hardware interface.
4. In the **hardware** tag, replace the existing plugin with the lightweight interface plugin shown below.

> ```xml

Reviewer 2: we need this back in. Sorry, my bad to an extent, I just couldn't comment a suggestion and also keep this in 😋
